hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManagerSyncJoinStrategy.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.cluster.impl;
+
+import com.hazelcast.internal.cluster.MemberInfo;
+import com.hazelcast.internal.cluster.impl.operations.OnJoinOp;
+import com.hazelcast.internal.util.BiTuple;
+import com.hazelcast.internal.util.Clock;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.spi.properties.ClusterProperty;
+import static java.lang.String.format;
+
+/**
+ * mimics functionality when join is delayed and clients are blocked until a
+ * certain timeout
+ * Can be removed when this strategy is obsoleted
+ *
+ * @author lprimak
+ */
+public class ClusterJoinManagerSyncJoinStrategy {
+    long timeToStartJoin;
+    private final ILogger logger;
+    private final long maxWaitMillisBeforeJoin;
+    private final long waitMillisBeforeJoin;
+
+    private long firstJoinRequest;
+
+    ClusterJoinManagerSyncJoinStrategy(ILogger logger, long maxWaitMillisBeforeJoin,
+            long waitMillisBeforeJoin) {
+        this.logger = logger;
+        this.maxWaitMillisBeforeJoin = maxWaitMillisBeforeJoin;
+        this.waitMillisBeforeJoin = waitMillisBeforeJoin;
+    }
+
+    /**
+     * Start processing the join request. This method is executed by the master node. In the case that there hasn't been any
+     * previous join requests from the {@code memberInfo}'s address the master will first respond by sending the master answer.
+     *
+     * Also, during the first {@link ClusterProperty#MAX_WAIT_SECONDS_BEFORE_JOIN} period since the master received the first
+     * join request from any node, the master will always wait for {@link ClusterProperty#WAIT_SECONDS_BEFORE_JOIN} before
+     * allowing any join request to proceed. This means that in the initial period from receiving the first ever join request,
+     * every new join request from a different address will prolong the wait time. After the initial period, join requests
+     * will get processed as they arrive for the first time.
+     *
+     * @param memberInfo the joining member info
+     * @param preJoinOperation which is prepared on joining members and will run on the master
+     */
+    void startJoinRequest(ClusterJoinManager manager, MemberInfo memberInfo, OnJoinOp preJoinOperation) {
```

Comment:
Nit-pick: This class is always created once for our `ClusterJoinManager` (of which only 1 exists), so we could probably pass this to the class's constructor instead of as a method parameter each time.

Reply:
ok

Reply:
done